Replace t5 Radius Matching with DNN + Pairwise Hit Differences + Make DNN First t5 Cut

GNiendorf opened the pull request:
Slides with PR Overview:  [PR #103 Overview.pdf](https://github.com/user-attachments/files/17382152/PR.103.Overview.pdf)

This PR Final Timing
<img width="1406" alt="Screenshot 2024-10-14 at 2 09 04 PM" src="https://github.com/user-attachments/assets/d319df1e-bbb0-4c84-9f6a-e0940bc87bae">

Current Master Timing
<img width="1410" alt="372299297-a8a62749-1643-4540-b667-e23c7fc28d42" src="https://github.com/user-attachments/assets/77ba1c23-ee57-4978-9329-4e1f3a379c97">

Replaces the t5 radius matching cuts with the DNN. As mentioned below, I kept the signal efficiency constant at 97.5% in each eta bin, so because I removed a cut with non-unity signal eff it makes sense that the efficiency increases slightly along with the FR. Marking this PR as ready for review.

Edit:

Pairwise hit differences help to decorrelate the input variables, speeding up training time by a factor of 4-5. Final AUC is largely the same, but faster training and less correlation among input features is useful.

<img width="790" alt="Screenshot 2024-10-11 at 10 27 23 AM" src="https://github.com/user-attachments/assets/aa58b6b2-4a9b-4537-b8d7-b80705e0da91">

Edit 2:

Using pragma unrolls results in a significant speedup of the DNN, resulting in an overall timing improvement from placing the DNN as the first t5 cut.

No unrolls, DNN before all cuts
<img width="1406" alt="Screenshot 2024-10-12 at 12 15 30 PM" src="https://github.com/user-attachments/assets/666bbdbb-de7c-4e86-b7b7-152151b2b304">

With unrolls, DNN before all cuts
<img width="1406" alt="Screenshot 2024-10-12 at 12 17 35 PM" src="https://github.com/user-attachments/assets/98102afd-4498-46c7-bfd6-226fdf1dd674">

This PR Timing, prior to the unrolls optimization + moving DNN to front
<img width="1406" alt="Screenshot 2024-10-12 at 12 06 20 PM" src="https://github.com/user-attachments/assets/dcdd1892-d714-4583-8f7d-8efd60d129f2">

Current Master Timing (taken from above):
<img width="1410" alt="372299297-a8a62749-1643-4540-b667-e23c7fc28d42" src="https://github.com/user-attachments/assets/77ba1c23-ee57-4978-9329-4e1f3a379c97">

GNiendorf commented:
/run all

GNiendorf commented:
/run all

GNiendorf commented:
I kept the signal efficiency constant, so the efficiency increases slightly along with the FR. Marking this PR as ready for review.

GNiendorf commented:
Pairwise hit differences help to decorrelate the input variables, speeding up training time by a factor of 4-5. Final AUC is largely the same, but faster training and less correlation among input features is useful.

<img width="790" alt="Screenshot 2024-10-11 at 10 27 23 AM" src="https://github.com/user-attachments/assets/aa58b6b2-4a9b-4537-b8d7-b80705e0da91">


/run all

slava77 commented:
> I kept the signal efficiency constant, so the efficiency increases slightly along with the FR. Marking this PR as ready for review.

"constant" in which direction, vs eta or vs baseline?

GNiendorf commented:
> > I kept the signal efficiency constant, so the efficiency increases slightly along with the FR. Marking this PR as ready for review.
> 
> "constant" in which direction, vs eta or vs baseline?

Sorry, I mean constant in eta (so I kept it at 97.5% signal efficiency in each bin). So because I removed a cut which had some non-unity signal efficiency, it makes sense that the overall t5/pt5 efficiency increase slightly along with the FR.

GNiendorf commented:
Thank you for the comments Slava, working on them now. One additional optimization I just noticed is to use the eta value of one of the hits instead of the "t5_eta" variable since it was removed from the DNN, but I'll leave that to the next PR.

GNiendorf commented:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#pragma-unroll

I was playing around with some optimizations, and it seems like using pragma unroll in the linear layer and relu functions gives a small benefit? I'll leave it out of this PR for simplicity, but it might become useful later on.

<img width="1406" alt="Screenshot 2024-10-12 at 12 01 09 PM" src="https://github.com/user-attachments/assets/87e587e0-593a-4823-8257-158187466b19">

GNiendorf commented:
One interesting thing: If I put the DNN inference before the rz chi-squared and runQuintupletAlgoSelector cuts, I can actually get significantly faster performance with the unrolls @slava77 :

No unrolls
<img width="1406" alt="Screenshot 2024-10-12 at 12 15 30 PM" src="https://github.com/user-attachments/assets/666bbdbb-de7c-4e86-b7b7-152151b2b304">

With unrolls
<img width="1406" alt="Screenshot 2024-10-12 at 12 17 35 PM" src="https://github.com/user-attachments/assets/98102afd-4498-46c7-bfd6-226fdf1dd674">

GNiendorf commented:
I see a similar improvement on the L4 (above is L40)

Current Timing
<img width="1406" alt="Screenshot 2024-10-12 at 9 08 52 PM" src="https://github.com/user-attachments/assets/a7647cb6-39bb-4386-acc2-01a9983fbc96">

With unrolls, DNN before all cuts
<img width="1406" alt="Screenshot 2024-10-12 at 9 06 23 PM" src="https://github.com/user-attachments/assets/cc5893b3-fe94-4a03-83e8-60f8897ca91e">

GNiendorf commented:
Given the above, I think it makes sense to add the pragma's to this PR. Let me know if you think otherwise though @slava77 

/run all

slava77 commented on RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:
```diff
@@ -29,8 +30,10 @@ namespace lst::t5dnn {
+#pragma unroll OUT_FEATURES
     for (unsigned int i = 0; i < OUT_FEATURES; ++i) {
       output[i] = biases[i];
+#pragma unroll IN_FEATURES
```
are both similarly useful? I'd think the innermost is more significant

GNiendorf replied:
It seems the benefit is coming from the outermost pragma mainly.

Only innermost pragma
<img width="1406" alt="Screenshot 2024-10-14 at 8 17 31 AM" src="https://github.com/user-attachments/assets/e73309a0-fb67-4e6b-a569-c59762724a56">

Only outermost pragma
<img width="1406" alt="Screenshot 2024-10-14 at 8 22 06 AM" src="https://github.com/user-attachments/assets/9f73fc76-d109-4773-b722-a8aa98c32abb">

GNiendorf commented on RecoTracker/LSTCore/standalone/analysis/DNN/train_T5_DNN.ipynb:
```diff
@@ -714,6 +700,163 @@
+    "    \n",
+    "    # Print the cut values\n",
+    "    print(f\"pt: {pt_min} to {pt_max}\")\n",
+    "    print(\"97.5% Retention Cut:\", np.round(cut_values_97, 4), \" Mean: \", np.round(np.mean(cut_values_97),4))\n",
```
Fixed, thanks.

GNiendorf commented:
Changed the cut bins to use the eta value of the first hit instead of t5_eta.

/run all

slava77 commented on RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:
```diff
@@ -39,102 +39,88 @@ namespace lst::t5dnn {
 
   template <typename TAcc>
   ALPAKA_FN_ACC ALPAKA_FN_INLINE float runInference(TAcc const& acc,
```
```suggestion
  ALPAKA_FN_ACC ALPAKA_FN_INLINE bool runInference(TAcc const& acc,
```
to match what's actually returned

slava77 commented:
/run checks

slava77 commented on RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:
```diff
@@ -39,102 +39,88 @@ namespace lst::t5dnn {
+        r1 / kR_max,       // inner T3: First hit r normalized
 
-    float t5_eta = mdsInGPU.anchorEta[md_idx_for_t5_eta_phi];
+        eta2 - eta1,         // inner T3: Difference in eta between hit 2 and 1
+        (z2 - z1) / kZ_max,  // inner T3: Difference in z between hit 2 and 1 normalized
```
here and below, are we losing performance around eta or z around zero by taking deltas of abs values? Perhaps deltas of signed values are more useful?

GNiendorf replied:
I just gave this a try but didn't see any improvement. Marking as resolved.

slava77 commented on RecoTracker/LSTCore/src/alpaka/Quintuplet.h:
```diff
@@ -2383,83 +1989,17 @@ namespace lst {
-      return false;
+    // 5 categories for sigmas
+    float sigmas2[5], delta1[5], delta2[5], slopes[5];
+    bool isFlat[5];
```
is it 5->`Params_T5::kLayers` here? or is it some different 5?
perhaps xVec, yVec can be explicitly sized as well

GNiendorf commented:
It seems like only the final cut in these functions is contributing anything, I don't see any differences in performance from removing all but the dbeta cuts.

Timing Before Removing
<img width="1407" alt="Screenshot 2024-10-19 at 3 23 00 PM" src="https://github.com/user-attachments/assets/af582399-54cd-4207-9b7e-a215bf3848e9">

Timing After Removing
<img width="1407" alt="Screenshot 2024-10-19 at 3 19 33 PM" src="https://github.com/user-attachments/assets/d7df1c8f-bd76-40a3-9fb1-2a50935e743e">

GNiendorf commented:
/run all

GNiendorf commented:
@slava77 Given that only the dBeta and rz chi-squared cuts are left, i think it makes sense to leave this PR open to see if I can integrate those into the DNN before merging.

slava77 commented:
/run all

slava77 commented:
>  Run Standalone PU200 — Status: failure 

looks like due to rate limit https://www.docker.com/increase-rate-limits/

@ariostas 
is it possible that the 200/6 hours limit also has a lower period limit, like 1/minute to make sense out of these failures?
Perhaps some 30 sec sleep is needed in the CI

slava77 commented:
/run standalone

ariostas commented:
> is it possible that the 200/6 hours limit also has a lower period limit, like 1/minute to make sense out of these failures?
> Perhaps some 30 sec sleep is needed in the CI

I'm not sure. I've always interpreted that as a client-side limitation. I'm guessing that that particular machine that was running the job happened to pull a bunch of Docker images recently

slava77 commented:
> There was a problem while building and running in standalone mode. The logs can be found [here](https://github.com/SegmentLinking/cmssw/actions/runs/11523134527).

```
make -C code/rooutil/
...
cxxopts.h:241:46: error: no match for 'operator+' (operand types are 'const char8_t [10]' and 'const std::string' {aka 'const std::__cxx11::basic_string<char>'})
  241 |         : OptionParseException(u8"Argument " + LQUOTE + text + RQUOTE + u8" starts with a - but has incorrect syntax") {}
```

is there a special symbol somewhere in RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc ?

slava77 commented:
> > There was a problem while building and running in standalone mode. The logs can be found [here](https://github.com/SegmentLinking/cmssw/actions/runs/11523134527).
> 
> ```
> make -C code/rooutil/
> ...
> cxxopts.h:241:46: error: no match for 'operator+' (operand types are 'const char8_t [10]' and 'const std::string' {aka 'const std::__cxx11::basic_string<char>'})
>   241 |         : OptionParseException(u8"Argument " + LQUOTE + text + RQUOTE + u8" starts with a - but has incorrect syntax") {}
> ```
> 
> is there a special symbol somewhere in RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc ?

..., this part is not compiling write_lst_ntuple; seems like some kind of glitch or is the CI variant having an issue

slava77 commented on RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h:
```diff
@@ -2,18 +2,16 @@
 
   template <int FEATURES>
   ALPAKA_FN_ACC ALPAKA_FN_INLINE void relu_activation(float (&input)[FEATURES]) {
+#pragma unroll FEATURES
```
@GNiendorf 
do you see these warnings of ignored pragma in your tests as well?

GNiendorf replied:
Yeah it looks like the CPU backend ignores the pragmas, I just tested it and they don't give a speedup on CPU.

GNiendorf replied:
Maybe I should wrap #pragma unroll in an ifdef only for the CUDA backend? I was told by Dan that CMSSW might have their own version of this as well?

GNiendorf commented:
> > is there a special symbol somewhere in RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc ?
> > ..., this part is not compiling write_lst_ntuple; seems like some kind of glitch or is the CI variant having an issue
> 
> Oh this issue came up when we moved to C++20. I thought I had already fixed this in the devel branch

I haven't rebased my branch in the two weeks this PR has been open. Maybe that will fix things? Did you move the devel branch to C++ in that timeframe?

ariostas commented:
> I haven't rebased my branch in the two weeks this PR has been open. Maybe that will fix things? Did you move the devel branch to C++20 in that timeframe?

The PR merges before testing, so it must be that the devel branch hasn't been fixed.

ariostas commented:
Hmm it's pretty weird because in https://github.com/SegmentLinking/cmssw/commit/1e6527f4ff98ed5d83469c7a378420bd47ee700f I did fix it for the devel branch. I wonder if it's not doing the right merge.

ariostas commented:
I'm very confused, I'll just try again to see if it works.

/run standalone

ariostas commented:
It doesn't make any sense. Here's part of the error:

```
cxxopts.h: In constructor 'cxxopts::Option_syntax_exception::Option_syntax_exception(const std::string&)':
cxxopts.h:241:46: error: no match for 'operator+' (operand types are 'const char8_t [10]' and 'const std::string' {aka 'const std::__cxx11::basic_string<char>'})
  241 |         : OptionParseException(u8"Argument " + LQUOTE + text + RQUOTE + u8" starts with a - but has incorrect syntax") {}
      |                                ~~~~~~~~~~~~~ ^ ~~~~~~
      |                                |               |
      |                                |               const std::string {aka const std::__cxx11::basic_string<char>}
      |                                const char8_t [10]
```

And here's the actual file in this branch.

https://github.com/SegmentLinking/cmssw/blob/782cf2fadc77b31b757c46124f4bc1a69451c52a/RecoTracker/LSTCore/standalone/code/rooutil/cxxopts.h#L238-L243

The `u8` is not there. However, it's showing that `ignoring '#pragma unroll FEATURES'` warning, so it's definitely using the right branch. I wonder if it's using some cached file or something.

slava77 commented:
> I wonder if it's using some cached file or something.

> And here's the actual file in this branch.
> 
> [cmssw/RecoTracker/LSTCore/standalone/code/rooutil/cxxopts.h](https://github.com/SegmentLinking/cmssw/blob/782cf2fadc77b31b757c46124f4bc1a69451c52a/RecoTracker/LSTCore/standalone/code/rooutil/cxxopts.h#L238-L243)

ah, but that's not the actual file, not this PR 782cf2fadc77b3, it should be the target branch, since this PR doesn't touch it: CMSSW_14_1_0_pre3_LST_X_LSTCore_realfiles_batch1_devel ends on 7a2c4228f537542fb2ea9497fb214b732c0bc245
https://github.com/SegmentLinking/cmssw/blob/7a2c4228f537542fb2ea9497fb214b732c0bc245/RecoTracker/LSTCore/standalone/code/rooutil/cxxopts.h#L239-L242

the change came from 992947d8d5a766089dd9e088c184bf71e95958ca

slava77 commented:
/run checks
/run standalone
